packages/hint/src/lib/engine.ts

```diff
@@ -189,20 +207,20 @@ export class Engine extends EventEmitter {
         });
     }
 
-    public onHintEvent(id: string, eventName: string, listener: Function) {
+    public onHintEvent<K extends StringKeyOf<E>>(id: string, eventName: K, listener: (data: E[K], event: string) => void) {
         const that = this;
 
-        const createEventHandler = (handler: Function, hintId: string) => {
+        const createEventHandler = (handler: (data: E[K], event: string) => void, hintId: string) => {
             /*
              * Using fake `this` parameter for typing: https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters
              * `this.event` defined by eventemitter2: https://github.com/EventEmitter2/EventEmitter2#differences-non-breaking-compatible-with-existing-eventemitter
              */
-            return function (this: { event: string }, event: Event): Promise<any> | null {
+            return function (this: { event: string }, event: E[K]): Promise<any> | null {
                 const urlsIgnoredForAll = that.ignoredUrls.get('all') || [];
                 const urlsIgnoredForHint = that.ignoredUrls.get(hintId) || [];
                 const urlsIgnored = urlsIgnoredForHint.concat(urlsIgnoredForAll);
 
-                if (that.isIgnored(urlsIgnored, event.resource)) {
+                if (that.isIgnored(urlsIgnored, (event as any).resource)) {
```

Comment:
I wish. TypeScript (correctly) doesn't like that cast since `event` may be `Problem[]` (due to the `'print'` event). Since `isIgnored` can handle `undefined` I felt the cast to `any` would suffice.

A better refactor would eliminate `Problem[]` and replace it with something inheriting from `Event` with a `problems: Problem[]` member so we could define `event` as `extends Event` in the first place, but that was an extra road I didn't want to go down for this PR.

Reply:
ahhh, right, I forgot about the 'print' event.

> A better refactor would eliminate Problem[] and replace it with something inheriting from Event with a problems: Problem[] member so we could define event as extends Event in the first place, but that was an extra road I didn't want to go down for this PR.

Sounds good to me. Can you open an issue for that?

Reply:
Thanks!